presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -87,75 +63,146 @@ public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> cluste
+                                            + CassandraCqlUtils.cqlValue(toCQLCompatibleString(range.getHigh().getValue()),
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                case BELOW:
+                                    rangeConjuncts.add(CassandraCqlUtils.validColumnName(columnHandle.getName()) + " < "
+                                            + CassandraCqlUtils.cqlValue(toCQLCompatibleString(range.getHigh().getValue()),
+                                            columnHandle.getCassandraType()));
+                                    break;
+                                default:
+                                    throw new AssertionError("Unhandled bound: " + range.getHigh().getBound());
+                            }
                         }
-                        return ImmutableSet.of();
-                    },
-                    allOrNone -> ImmutableSet.of());
+                    }
+                }
 
-            if (!values.isEmpty()) {
-                clusteringColumnValues.add(values);
+                if (singleValues.size() == 1) {
+                    predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " = " + singleValues.get(0);
+                }
+                else if (singleValues.size() > 1) {
+                    if (!isPredicateSupported(cassandraVersion)) {
+                        return "Not Supported";
+                    }
+                    predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                            + Joiner.on(",").join(singleValues) + ")";
+                }
+                else if (!rangeConjuncts.isEmpty()) {
+                    predicate = Joiner.on(" AND ").join(rangeConjuncts);
+                }
+                return predicate;
+            }, discreteValues -> {
+                if (discreteValues.isWhiteList()) {
+                    ImmutableList.Builder<Object> discreteValuesList = ImmutableList.builder();
+                    if (!isPredicateSupported(cassandraVersion)) {
+                        return "Not Supported";
+                    }
+                    for (Object discreteValue : discreteValues.getValues()) {
+                        discreteValuesList.add(CassandraCqlUtils.cqlValue(toCQLCompatibleString(discreteValue),
+                                columnHandle.getCassandraType()));
+                    }
+                    String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
+                            + Joiner.on(",").join(discreteValuesList.build()) + ")";
+                    return predicate;
+                }
+                return null;
+            }, allOrNone -> null);
+
+            if (predicateString != null) {
+                // Check for Cassandra version older than 2.2.x
```

Comment:
This comment isn't consistent with the version 2.1 that you have in `isPredicateSupported`.